File: Kirameku-backend/admin/src/api/secretCode.ts
import { http } from "@/utils/http";

export type SecretCodeTargetType = "internal" | "external" | "download";

export type SecretCodeItem = {
  id: number;
  name: string;
  code: string;
  description: string;
  target_type: SecretCodeTargetType;
  target_url: string;
  is_active: boolean;
  expires_at: string | null;
  max_uses: number;
  use_count: number;
  last_used_at: string | null;
  created_at: string;
  updated_at: string;
};

export type SecretCodePayload = Pick<
  SecretCodeItem,
  | "name"
  | "code"
  | "description"
  | "target_type"
  | "target_url"
  | "is_active"
  | "expires_at"
  | "max_uses"
>;

export const getSecretCodes = () => {
  return http.request<SecretCodeItem[]>("get", "/api/secret-codes/admin");
};

export const createSecretCode = (data: SecretCodePayload) => {
  return http.request<SecretCodeItem>("post", "/api/secret-codes", { data });
};

export const updateSecretCode = (
  codeId: number,
  data: Partial<SecretCodePayload>
) => {
  return http.request<SecretCodeItem>("put", `/api/secret-codes/${codeId}`, {
    data
  });
};

export const resetSecretCodeUses = (codeId: number) => {
  return http.request<SecretCodeItem>(
    "post",
    `/api/secret-codes/${codeId}/reset-uses`
  );
};

export const deleteSecretCode = (codeId: number) => {
  return http.request<{ ok: boolean }>("delete", `/api/secret-codes/${codeId}`);
};

File: Kirameku-backend/app/models/secret_code.py
from datetime import datetime
from typing import Optional

from sqlmodel import Field, SQLModel


class SecretCode(SQLModel, table=True):
    __tablename__ = "secret_code"

    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(max_length=100)
    code: str = Field(max_length=100, unique=True, index=True)
    description: str = Field(default="", max_length=500)
    target_type: str = Field(default="internal", max_length=20)
    target_url: str = Field(max_length=2000)
    is_active: bool = Field(default=True)
    expires_at: Optional[datetime] = Field(default=None)
    max_uses: int = Field(default=0)
    use_count: int = Field(default=0)
    last_used_at: Optional[datetime] = Field(default=None)
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
